x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/persistence/JobResultsProvider.java

```diff
@@ -442,6 +444,51 @@ private SearchRequestBuilder createTimingStatsSearch(String indexName, String jo
             .setQuery(QueryBuilders.idsQuery().addIds(TimingStats.documentId(jobId)));
     }
 
+    public void datafeedTimingStats(Collection<String> jobIds, Consumer<Map<String, DatafeedTimingStats>> handler,
+                                    Consumer<Exception> errorHandler) {
+        String indexName = AnomalyDetectorsIndex.jobResultsIndexName();
+        SearchRequest searchRequest =
+            client.prepareSearch(indexName)
+                .setSize(jobIds.size())
```

Comment:
This only works if all the datafeed timing stats can be guaranteed to be in the same index.  But as I said in a previous comment I don't think it's a good idea to introduce a result document that always gets written to `.ml-anomalies-shared` even if every other result document for the job is going somewhere else.

But, as you may have realised, there is a complication if we write the datafeed stats documents to the current write alias for each job: if that alias switches to a different index then the more recent datafeed stats document for that job will not overwrite the older document(s).  So I think you should introduce a method to search for the latest datafeed stats for one job, using `setSize(1)` and a descending sort on total search time.  This is similar to how we find the latest data counts for a job.  Then this method that gets the stats for many jobs will have to loop and call the single job method once per job.  I agree this is less efficient than searching for many documents in the same search, but it's what we do for data counts so it cannot cause that many problems in practice and it is robust to job results indices rolling over.